### ISSUE TITLE
reproduce rpg and rmpg on kuhn and leduc

### ISSUE THREAD

**Issue by lihuiknight** [deepmind/open_spiel#155]
Dear Authors,

I am trying to reproduce the results of https://arxiv.org/pdf/1810.09026.pdf based on pytorch. The idea in this paper is very interesting. Before my implementation, I run the experiments on two-players kuhn and leduc based on the release codes: [kuhn_nfsp.py](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/examples/kuhn_nfsp.py) and [kuhn_policy_gradient.py](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/examples/kuhn_policy_gradient.py).

To see the long-term convergence, I set num_episodes=1e7 and use the default settings for other hyperparameters,  the exploitability of different methods after 1e7 episode are listed as follows:
          a2c, qpg, rpg, rmpg, nfsp
kuhn:  0.19757, 0.19307, 0.19336, 0.12419, 0.01980
leduc: 1.05118, 0.60193, 0.67432, 1.22162, 0.63754

It seems that the results are worse than the results reported in Figure 2. I wonder whether it's due to the high well-known variance in deep rl, the fine-tuning hyperparameters or some other tricks I have missed.

Best,
Hui

**Comment by lanctot** [comment]
Hi @lihuiknight,

Ok yes, several things:

1. Please do not refer to the original paper for results, but rather the [OpenSpiel paper](https://arxiv.org/abs/1908.09453) (Figure 3, page 17). In the original paper, we used a more difficult version of Leduc which, rather than handling illegal actions explicitly, imposed arbitrary penalties when taking illegal actions (and then sampled uniformly among legal actions). In the OpenSpiel paper, @fzvinicius reproduced our curves on Leduc but with the OpenSpiel version of the algorithms, which had the standard version of Leduc. I will ask him to post the hyper-parameters he used for these runs.

2. We found that entropy bonuses to be important for good convergence performatce (and also updating the value heads more often than the policy heads). I don't remember what we included in those examples, but probably just a simple use of the policy gradient agents. I will let Vinicius follow-up on what his experiment did to produce the values in Figure 3.

**Comment by lanctot** [comment]
(Note that Vinicius is currently on holiday so he might not respond right away. I will remind him to take a look when he's back.)

**Comment by lihuiknight** [comment]
Thanks @lanctot . It's expected that deep rl algorithm without correct hyperparameters or other tricks will not work well, as discussed in https://arxiv.org/pdf/1709.06560.pdf.

According to the implementation in open_spiel, `entropy_cost=0.01, num_critic_before_pi=8,`. Using entropy bonuses and updating value network more frequent than policy network are two important tricks in rl community.

Looking forward to hearing from Vinicius.

--------------------------------------------------------------------------

Also, I have another question: 
in the implementation,  the on-policy regret gradient descent using mini-batch stochastic gradient descent (batch-size=128) is very confusing. I think it will be a biased estimate. Why not use important sampling?

**Comment by lanctot** [comment]
> in the implementation, the on-policy regret gradient descent using mini-batch stochastic gradient descent (batch-size=128) is very confusing. I think it will be a biased estimate. Why not use important sampling?

You mean this? https://github.com/deepmind/open_spiel/blob/ed118af36137c1b01b3d54502747eb1ac28d939a/open_spiel/python/algorithms/policy_gradient.py#L103

I might not understand your question.. why would it be biased (and why would the size of the batch matter for determining that?) The learning is done synchronously (as in batch A2C) so there is no lag in the learning and acting here: data is collected and only the collected data is used to update the critics and policy (it's thrown away after the learning step here: https://github.com/deepmind/open_spiel/blob/ed118af36137c1b01b3d54502747eb1ac28d939a/open_spiel/python/algorithms/policy_gradient.py#L305). So nothing is off-policy, so I don't understand why you would need importance sampling?

**Comment by lihuiknight** [comment]
@lanctot  Yes. You're correct. During the batch training, the policy is not updated.  Because the self._dataset has been thrown away, it doesn't need importance sampling. If this dataset is maintained in a buffer, i think importance sampling should be used here.

**Comment by fzvinicius** [comment]
Hi @lihuiknight ,

Thank you for bearing with me on this query.
As you mentioned, there is a good amount of variance in deep RL. In general this effect is even more pronounced on PG algorithms, which is probably why the values you obtained are not aligned with the paper.

I compiled some of the hyperparameters we used for the paper on the table below:

![image](https://user-images.githubusercontent.com/194721/77430244-c2fad680-6dd2-11ea-854f-60cd954323f1.png)

Please also remember to set ```num_episodes = 2e7```.

**Comment by lihuiknight** [comment]
OK. Thanks. I close this issue.
